src/OrleansRuntime/LogViews/PrimaryBasedLogViewAdaptor.cs

```diff
@@ -0,0 +1,840 @@
+            }
+        }
+
+        /// <summary>
+        /// Merge two notification messages, for batching. Override to handle notification subtypes.
+        /// </summary>
+        protected virtual INotificationMessage Merge(INotificationMessage earliermessage, INotificationMessage latermessage)
+        {
+            return new VersionNotificationMessage()
+            {
+                Version = latermessage.Version
+            };
+        }
+
+        /// <summary>
+        /// Called when configuration of the multicluster is changing.
+        /// </summary>
+        protected virtual Task OnConfigurationChange(MultiClusterConfiguration next)
+        {
+            Configuration = next;
+            return TaskDone.Done;
+        }
+
+        /// <summary>
+        /// The grain that is using this adaptor
+        /// </summary>
+        protected ILogViewHost<TLogView, TLogEntry> Host { get; private set; }
+
+        protected IProtocolServices Services { get; private set; }
+
+        protected MultiClusterConfiguration Configuration { get; set; }
+
+        protected ILogViewProvider Provider;
+
+        /// <summary>
+        /// Tracks notifications sent. Created lazily since many copies will never need to send notifications.
+        /// </summary>
+        private NotificationTracker notificationtracker;
+
+
+        private const int max_notification_batch_size = 10000;
+
+
+        protected PrimaryBasedLogViewAdaptor(ILogViewHost<TLogView, TLogEntry> host, ILogViewProvider provider,
+            TLogView initialstate, IProtocolServices services)
+        {
+            Debug.Assert(host != null && services != null && initialstate != null);
+            this.Host = host;
+            this.Services = services;
+            this.Provider = provider;
+            InitializeConfirmedView(initialstate);
+            worker = new BatchWorkerFromDelegate(() => Work());
+        }
+
+        public virtual async Task Activate()
+        {
+            Services.Verbose2("Activation Started");
+
+            if (Silo.CurrentSilo.GlobalConfig.HasMultiClusterNetwork)
+            {
+                // subscribe this grain to configuration change events
+                Silo.CurrentSilo.LocalMultiClusterOracle.SubscribeToMultiClusterConfigurationEvents(Services.GrainReference);
+            }
+
+            // initial load happens async
+            KickOffInitialRead().Ignore();
+
+            var latestconf = Services.MultiClusterConfiguration;
+            if (latestconf != null)
+                await OnMultiClusterConfigurationChange(latestconf);
+
+            Services.Verbose2("Activation Complete");
+        }
+
+        private async Task KickOffInitialRead()
+        {
+            need_initial_read = true;
+            // kick off notification for initial read cycle with a bit of delay
+            // so that we don't do this several times if user does strong sync
+            await Task.Delay(10);
```

Comment:
Maybe a matter of preference and especially now that this is an "arbitrary technical delay", but using explicit `TimeSpan` would look more robust.